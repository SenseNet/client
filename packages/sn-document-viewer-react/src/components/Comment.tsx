import Avatar from '@material-ui/core/Avatar'
import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import CardHeader from '@material-ui/core/CardHeader'
import Collapse from '@material-ui/core/Collapse'
import React, { useState } from 'react'
import { FunctionComponent } from 'react'
import { connect } from 'react-redux'
import styled from 'styled-components'
import { Comment as CommentType } from '../models/Comment'
import { componentType } from '../services'
import { deleteComment, RootReducerType } from '../store'

const mapStateToProps = (state: RootReducerType) => ({
  localization: state.sensenetDocumentViewer.localization,
})

const mapDispatchToProps = {
  deleteComment,
}

const StyledCardContent = styled(CardContent)`
  font-size: 14px;
  letter-spacing: 0.2px;
  line-height: 20px;
  word-wrap: break-word;
`
const DeleteButton = (props: CommentPropType) => (
  <Button size="small" onClick={() => props.deleteComment(props.id)}>
    {props.localization.delete || 'delete'}
  </Button>
)

/**
 * Comment prop type
 */
export type CommentPropType = componentType<typeof mapStateToProps, typeof mapDispatchToProps, CommentType>

const MAX_TEXT_LENGTH = 160

/**
 * Represents a single comment component.
 */
export const CommentComponent: FunctionComponent<CommentPropType> = props => {
  const isLongText = props.text && props.text.length > MAX_TEXT_LENGTH
  const [isOpen, setIsOpen] = useState(!isLongText)
  return (
    <Card>
      <CardHeader
        avatar={<Avatar src={props.createdBy.avatarUrl} alt={props.localization.avatarAlt} />}
        title={props.createdBy.displayName}
      />
      <Collapse in={isOpen} timeout="auto" collapsedHeight={isOpen ? '0px' : '78px'}>
        <StyledCardContent>{props.text}</StyledCardContent>
      </Collapse>
      <CardActions>
        {isLongText ? (
          <>
            <Button size="small" onClick={() => setIsOpen(!isOpen)}>
              {isOpen ? props.localization.showLess || 'Show less' : props.localization.showMore || 'Show more'}
            </Button>
            {isOpen ? <DeleteButton {...props} /> : null}
          </>
        ) : (
          <DeleteButton {...props} />
        )}
      </CardActions>
    </Card>
  )
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CommentComponent)
