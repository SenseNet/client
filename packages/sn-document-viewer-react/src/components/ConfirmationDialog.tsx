import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogTitle from '@material-ui/core/DialogTitle'
import React from 'react'
import { FunctionComponent } from 'react'

/**
 * Interface for confirmation dialog component properties
 */
export interface ConfirmationDialogProps {
  dialogTitle: string
  okButtonText?: string
  cancelButtonText?: string
  isOpen: boolean
  onClose: (isCanceled: boolean) => void
}

/**
 * Represents a confirmation dialog
 */
export const ConfirmationDialog: FunctionComponent<ConfirmationDialogProps> = props => {
  const handleCancel = () => props.onClose(true)
  const handleOk = () => props.onClose(false)
  return (
    <Dialog
      disableBackdropClick={true}
      disableEscapeKeyDown={true}
      maxWidth="md"
      aria-labelledby="confirmation-dialog-title"
      aria-describedby="confirmation-dialog-description"
      open={props.isOpen}>
      <DialogTitle id="confirmation-dialog-title">{props.dialogTitle}</DialogTitle>
      <DialogContent>
        <DialogContentText
          dangerouslySetInnerHTML={{ __html: props.children as string }}
          id="confirmation-dialog-description"
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={handleOk} variant="contained" color="primary">
          {props.okButtonText || 'ok'}
        </Button>
        <Button onClick={handleCancel}>{props.cancelButtonText || 'cancel'}</Button>
      </DialogActions>
    </Dialog>
  )
}
