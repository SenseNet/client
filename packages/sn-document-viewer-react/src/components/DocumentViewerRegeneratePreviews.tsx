import Typography from '@material-ui/core/Typography'
import React = require('react')
import { connect } from 'react-redux'
import { Button } from '@material-ui/core'
import { componentType } from '../services/TypeHelpers'
import { regeneratePreviews, RootReducerType } from '../store'
import { LayoutAppBar } from './LayoutAppBar'

/**
 * maps state fields from the store to component props
 * @param state the redux state
 */
const mapStateToProps = (state: RootReducerType) => {
  return {
    regeneratePreviewsText: state.sensenetDocumentViewer.localization.regeneratePreviews,
    regenerateButton: state.sensenetDocumentViewer.localization.regenerateButton,
  }
}

/**
 * maps state actions from the store to component props
 * @param state the redux state
 */
const mapDispatchToProps = {
  regeneratePreviews,
}

class DocumentViewerRegeneratePreviewsComponent extends React.Component<
  componentType<typeof mapStateToProps, typeof mapDispatchToProps>
> {
  /**
   * renders the component
   */
  public render() {
    return (
      <div
        style={{
          display: 'flex',
          justifyContent: 'space-around',
          alignContent: 'center',
          alignItems: 'center',
          height: '100%',
        }}>
        <LayoutAppBar style={{ position: 'fixed', top: 0 }}>
          <span />
        </LayoutAppBar>
        <div
          style={{
            display: 'flex',
            justifyContent: 'space-between',
            alignItems: 'center',
            flexDirection: 'column',
            maxWidth: 500,
            margin: '.5em 0 .6em 0',
          }}>
          <Typography variant="subtitle1" color="textSecondary" align="center" gutterBottom>
            {this.props.regeneratePreviewsText}
          </Typography>
          <Button variant="contained" onClick={this.props.regeneratePreviews}>
            {this.props.regenerateButton}
          </Button>
        </div>
      </div>
    )
  }
}

const connectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps,
)(DocumentViewerRegeneratePreviewsComponent)

export { connectedComponent as DocumentViewerRegeneratePreviews }
