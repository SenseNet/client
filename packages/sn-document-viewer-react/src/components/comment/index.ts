export * from './comment'
export * from './create-comment'
export * from './delete-comment-button'
export * from './style'
