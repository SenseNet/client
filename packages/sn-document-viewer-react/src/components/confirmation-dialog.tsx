import { Button, Dialog, DialogActions, DialogContent, DialogContentText, DialogTitle } from '@material-ui/core'
import React from 'react'

/**
 * Interface for confirmation dialog component properties
 */
export interface ConfirmationDialogProps {
  dialogTitle: string
  okButtonText?: string
  cancelButtonText?: string
  isOpen: boolean
  onClose: (isCanceled: boolean) => void
}

/**
 * Represents a confirmation dialog
 */
export const ConfirmationDialog: React.FC<ConfirmationDialogProps> = (props) => {
  const handleCancel = () => props.onClose(true)
  const handleOk = () => props.onClose(false)
  return (
    <Dialog
      disableBackdropClick={true}
      disableEscapeKeyDown={true}
      maxWidth="md"
      aria-labelledby="confirmation-dialog-title"
      aria-describedby="confirmation-dialog-description"
      open={props.isOpen}>
      <DialogTitle id="confirmation-dialog-title">{props.dialogTitle}</DialogTitle>
      <DialogContent>
        <DialogContentText
          dangerouslySetInnerHTML={{ __html: props.children as string }}
          id="confirmation-dialog-description"
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={handleOk} variant="contained" color="primary">
          {props.okButtonText || 'ok'}
        </Button>
        <Button onClick={handleCancel}>{props.cancelButtonText || 'cancel'}</Button>
      </DialogActions>
    </Dialog>
  )
}
