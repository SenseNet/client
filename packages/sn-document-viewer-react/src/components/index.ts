export * from './comment'
export * from './document-widgets'
export * from './page-widgets'
export * from './ConfirmationDialog'
export * from './DocumentViewer'
export * from './DocumentViewerError'
export * from './DocumentViewerLayout'
export * from './DocumentViewerLoading'
export * from './DocumentViewerRegeneratePreviews'
export * from './LayoutAppBar'
export * from './Page'
export * from './PageList'
export * from './styles'
