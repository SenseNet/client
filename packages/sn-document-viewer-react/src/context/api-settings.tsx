import React, { useEffect, useState } from 'react'
import { useRepository } from '@sensenet/hooks-react'
import { File as SnFile } from '@sensenet/default-content-types'
import { deepMerge, toNumber } from '@sensenet/client-utils'
import { v1 } from 'uuid'
import { Repository } from '@sensenet/client-core'
import {
  Annotation,
  CommentData,
  CommentWithoutCreatedByAndId,
  DocumentData,
  DocumentViewerApiSettings,
  Highlight,
  PreviewImageData,
  Redaction,
  Shape,
} from '../models'

/**
 * Adds a globally unique ID to the shape
 */
const addGuidToShape = <T extends Shape>(shape: T) => {
  shape.guid = v1()
  return shape
}

/**
 * Concatenates avatar url with host
 * @param {DocumentData} documentData
 */
function changeCreatedByUrlToCurrent(documentData: DocumentData): (value: CommentData) => CommentData {
  return (comment) => {
    return {
      ...comment,
      createdBy: { ...comment.createdBy, avatarUrl: `${documentData.hostName}${comment.createdBy.avatarUrl}` },
    }
  }
}

export const DocumentViewerApiSettingsContext = React.createContext<DocumentViewerApiSettings>({
  canEditDocument: async () => false,
  canHideRedaction: async () => false,
  canHideWatermark: async () => false,
  commentActions: {
    addPreviewComment: async ({ comment }) => ({ ...comment } as any),
    deletePreviewComment: async () => ({ modified: false }),
    getPreviewComments: async () => [],
  },
  getDocumentData: async () => ({
    documentName: 'example',
    documentType: '',
    fileSizekB: 3,
    hostName: '',
    idOrPath: 1,
    pageAttributes: [],
    pageCount: 0,
    shapes: { annotations: [], highlights: [], redactions: [] },
  }),
  getExistingPreviewImages: async () => [],
  isPreviewAvailable: async () => ({ Height: 0, Width: 0, Index: 0, Attributes: { degree: 0 } }),
  regeneratePreviews: async () => undefined,
  saveChanges: async () => undefined,
})

export const createDefaultApiSettings: (repo: Repository) => DocumentViewerApiSettings = (repo: Repository) => ({
  regeneratePreviews: async ({ document, abortController }) => {
    repo.executeAction({
      idOrPath: document.idOrPath,
      method: 'POST',
      body: undefined,
      name: 'RegeneratePreviews',
      requestInit: {
        signal: abortController.signal,
      },
    })
  },
  saveChanges: async ({ document, pages, abortController }) => {
    const reqBody = {
      Shapes: JSON.stringify([
        { redactions: document.shapes.redactions },
        { highlights: document.shapes.highlights },
        { annotations: document.shapes.annotations },
      ]),
      PageAttributes: JSON.stringify(
        pages
          .map((p) => (p.Attributes && p.Attributes.degree && { pageNum: p.Index, options: p.Attributes }) || undefined)
          .filter((p) => p !== undefined),
      ),
    }
    await repo.patch<SnFile>({
      idOrPath: document.idOrPath,
      content: reqBody,
      requestInit: {
        signal: abortController.signal,
      },
    })
  },
  getDocumentData: async (settings) => {
    const documentData = (
      await repo.load<SnFile>({
        idOrPath: settings.idOrPath,
        oDataOptions: {
          select: 'all',
        },
      })
    ).d

    return {
      idOrPath: settings.idOrPath,
      hostName: repo.configuration.repositoryUrl,
      fileSizekB: documentData.Size as number,
      pageCount: toNumber(documentData.PageCount, -1)!,
      documentName: documentData.DisplayName || '',
      documentType: documentData.Type || 'File',
      shapes: (documentData.Shapes && {
        redactions: (JSON.parse(documentData.Shapes)[0].redactions as Redaction[]).map((a) => addGuidToShape(a)) || [],
        annotations:
          (JSON.parse(documentData.Shapes)[2].annotations as Annotation[]).map((a) => addGuidToShape(a)) || [],
        highlights: (JSON.parse(documentData.Shapes)[1].highlights as Highlight[]).map((a) => addGuidToShape(a)) || [],
      }) || {
        redactions: [],
        annotations: [],
        highlights: [],
      },
      pageAttributes: (documentData.PageAttributes && JSON.parse(documentData.PageAttributes)) || [],
    }
  },
  isPreviewAvailable: async ({ document, version, page, abortController }) => {
    const responseBody = await repo.executeAction<{ page: number }, PreviewImageData & { PreviewAvailable: string }>({
      idOrPath: document.idOrPath,
      method: 'POST',
      name: `PreviewAvailable?version=${version}`,
      body: {
        page,
      },
      requestInit: {
        signal: abortController.signal,
      },
    })

    if (responseBody.PreviewAvailable) {
      responseBody.PreviewImageUrl = `${document.hostName}${responseBody.PreviewAvailable}`
      responseBody.ThumbnailImageUrl = `${document.hostName}${responseBody.PreviewAvailable.replace(
        'preview',
        'thumbnail',
      )}`
      return responseBody as PreviewImageData
    }
  },
  canEditDocument: async ({ document, abortController }) => {
    const response = await repo.security.hasPermission(document.idOrPath, ['Save'], undefined, {
      signal: abortController.signal,
      credentials: 'include',
    })
    return response
  },
  canHideRedaction: async ({ document, abortController }) =>
    await repo.security.hasPermission(document.idOrPath, ['PreviewWithoutRedaction'], undefined, {
      signal: abortController.signal,
      credentials: 'include',
    }),
  canHideWatermark: async ({ document, abortController }) =>
    await repo.security.hasPermission(document.idOrPath, ['PreviewWithoutWatermark'], undefined, {
      signal: abortController.signal,
      credentials: 'include',
    }),
  getExistingPreviewImages: async ({ document, version, abortController }) => {
    if (document.pageCount < -1) {
      throw Error('Preview generation error')
    }

    const response = await repo.executeAction({
      idOrPath: document.idOrPath,
      name: `GetExistingPreviewImages`,
      method: 'POST',
      body: {},
      oDataOptions: {
        select: 'all',
        expand: 'all',
        version,
      } as any,
      requestInit: {
        signal: abortController.signal,
      },
    })

    const availablePreviews = ((await response) as Array<PreviewImageData & { PreviewAvailable?: string }>).map((a) => {
      if (a.PreviewAvailable) {
        a.PreviewImageUrl = `${document.hostName}${a.PreviewAvailable}`
        a.ThumbnailImageUrl = `${document.hostName}${a.PreviewAvailable.replace('preview', 'thumbnail')}`
      }
      return a
    })

    const allPreviews: PreviewImageData[] = []
    for (let i = 0; i < document.pageCount; i++) {
      allPreviews[i] = availablePreviews[i] || ({ Index: i + 1 } as any)
      const pageAttributes = document.pageAttributes.find((p) => p.pageNum === allPreviews[i].Index)
      allPreviews[i].Attributes = pageAttributes && pageAttributes.options
    }
    return allPreviews
  },
  commentActions: {
    addPreviewComment: async ({ document, comment, abortController }) => {
      const response = await repo.executeAction<CommentWithoutCreatedByAndId, CommentData>({
        idOrPath: document.idOrPath,
        body: comment,
        name: 'AddPreviewComment',
        method: 'POST',
        requestInit: {
          signal: abortController.signal,
        },
      })
      return [response].map(changeCreatedByUrlToCurrent(document))[0]
    },
    deletePreviewComment: async ({ document, commentId, abortController }) => {
      const response = await repo.executeAction<any, { modified: boolean }>({
        idOrPath: document.idOrPath,
        body: { id: commentId },
        name: 'DeletePreviewComment',
        method: 'POST',
        requestInit: {
          signal: abortController.signal,
        },
      })
      return response
    },
    getPreviewComments: async ({ document, page, abortController }) => {
      const response = await repo.executeAction<any, CommentData[]>({
        idOrPath: document.idOrPath,
        name: 'GetPreviewComments',
        method: 'GET',
        oDataOptions: {
          page,
        } as any,
        requestInit: {
          signal: abortController.signal,
        },
      })
      return response.map(changeCreatedByUrlToCurrent(document))
    },
  },
})

export const DocumentViewerApiSettingsProvider: React.FC<{ options?: Partial<DocumentViewerApiSettings> }> = ({
  children,
  options,
}) => {
  const repo = useRepository()
  const [settings, setSettings] = useState(deepMerge(createDefaultApiSettings(repo), options))
  useEffect(() => {
    setSettings(deepMerge(createDefaultApiSettings(repo), options))
  }, [repo, options])

  return (
    <DocumentViewerApiSettingsContext.Provider value={settings}>{children}</DocumentViewerApiSettingsContext.Provider>
  )
}
