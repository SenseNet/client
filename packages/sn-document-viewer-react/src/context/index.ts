export * from './api-settings'
export * from './comment-states'
export * from './comments'
export * from './document-data'
export * from './document-permissions'
export * from './localization-context'
export * from './preview-image-data'
export * from './viewer-settings'
export * from './viewer-state'
