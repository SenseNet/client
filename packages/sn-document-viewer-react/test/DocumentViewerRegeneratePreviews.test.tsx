import { Button } from '@material-ui/core'
import { mount, shallow } from 'enzyme'
import React from 'react'
import { DocumentViewerRegeneratePreviews } from '../src/components/DocumentViewerRegeneratePreviews'
import { DocumentViewerApiSettingsContext } from '../src/context/api-settings'
import { defaultSettings } from './__Mocks__/viewercontext'

describe('Document Regenerate Previews component', () => {
  it('should render without crashing', () => {
    const wrapper = shallow(<DocumentViewerRegeneratePreviews />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Should trigger regeneration on button click', () => {
    const regeneratePreviews = jest.fn(async () => undefined)
    const wrapper = mount(
      <DocumentViewerApiSettingsContext.Provider value={{ ...defaultSettings, regeneratePreviews }}>
        <DocumentViewerRegeneratePreviews />
      </DocumentViewerApiSettingsContext.Provider>,
    )
    wrapper.find(Button).simulate('click')
    expect(regeneratePreviews).toBeCalled()

    expect(wrapper).toMatchSnapshot()
  })
})
