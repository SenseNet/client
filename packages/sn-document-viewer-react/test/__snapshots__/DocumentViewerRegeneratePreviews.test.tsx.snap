// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document Regenerate Previews component Should trigger regeneration on button click 1`] = `
<Component>
  <div
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "space-around",
      }
    }
  >
    <Component
      style={
        Object {
          "position": "fixed",
          "top": 0,
        }
      }
    >
      <WithStyles(ForwardRef(AppBar))
        position="sticky"
        style={
          Object {
            "position": "fixed",
            "top": 0,
            "zIndex": 1,
          }
        }
      >
        <ForwardRef(AppBar)
          classes={
            Object {
              "colorDefault": "MuiAppBar-colorDefault",
              "colorInherit": "MuiAppBar-colorInherit",
              "colorPrimary": "MuiAppBar-colorPrimary",
              "colorSecondary": "MuiAppBar-colorSecondary",
              "colorTransparent": "MuiAppBar-colorTransparent",
              "positionAbsolute": "MuiAppBar-positionAbsolute",
              "positionFixed": "MuiAppBar-positionFixed",
              "positionRelative": "MuiAppBar-positionRelative",
              "positionStatic": "MuiAppBar-positionStatic",
              "positionSticky": "MuiAppBar-positionSticky",
              "root": "MuiAppBar-root",
            }
          }
          position="sticky"
          style={
            Object {
              "position": "fixed",
              "top": 0,
              "zIndex": 1,
            }
          }
        >
          <WithStyles(ForwardRef(Paper))
            className="MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary"
            component="header"
            elevation={4}
            square={true}
            style={
              Object {
                "position": "fixed",
                "top": 0,
                "zIndex": 1,
              }
            }
          >
            <ForwardRef(Paper)
              className="MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary"
              classes={
                Object {
                  "elevation0": "MuiPaper-elevation0",
                  "elevation1": "MuiPaper-elevation1",
                  "elevation10": "MuiPaper-elevation10",
                  "elevation11": "MuiPaper-elevation11",
                  "elevation12": "MuiPaper-elevation12",
                  "elevation13": "MuiPaper-elevation13",
                  "elevation14": "MuiPaper-elevation14",
                  "elevation15": "MuiPaper-elevation15",
                  "elevation16": "MuiPaper-elevation16",
                  "elevation17": "MuiPaper-elevation17",
                  "elevation18": "MuiPaper-elevation18",
                  "elevation19": "MuiPaper-elevation19",
                  "elevation2": "MuiPaper-elevation2",
                  "elevation20": "MuiPaper-elevation20",
                  "elevation21": "MuiPaper-elevation21",
                  "elevation22": "MuiPaper-elevation22",
                  "elevation23": "MuiPaper-elevation23",
                  "elevation24": "MuiPaper-elevation24",
                  "elevation3": "MuiPaper-elevation3",
                  "elevation4": "MuiPaper-elevation4",
                  "elevation5": "MuiPaper-elevation5",
                  "elevation6": "MuiPaper-elevation6",
                  "elevation7": "MuiPaper-elevation7",
                  "elevation8": "MuiPaper-elevation8",
                  "elevation9": "MuiPaper-elevation9",
                  "outlined": "MuiPaper-outlined",
                  "root": "MuiPaper-root",
                  "rounded": "MuiPaper-rounded",
                }
              }
              component="header"
              elevation={4}
              square={true}
              style={
                Object {
                  "position": "fixed",
                  "top": 0,
                  "zIndex": 1,
                }
              }
            >
              <header
                className="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
                style={
                  Object {
                    "position": "fixed",
                    "top": 0,
                    "zIndex": 1,
                  }
                }
              >
                <WithStyles(ForwardRef(Toolbar))
                  style={
                    Object {
                      "display": "flex",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <ForwardRef(Toolbar)
                    classes={
                      Object {
                        "dense": "MuiToolbar-dense",
                        "gutters": "MuiToolbar-gutters",
                        "regular": "MuiToolbar-regular",
                        "root": "MuiToolbar-root",
                      }
                    }
                    style={
                      Object {
                        "display": "flex",
                        "justifyContent": "space-between",
                      }
                    }
                  >
                    <div
                      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
                      style={
                        Object {
                          "display": "flex",
                          "justifyContent": "space-between",
                        }
                      }
                    >
                      <span />
                    </div>
                  </ForwardRef(Toolbar)>
                </WithStyles(ForwardRef(Toolbar))>
              </header>
            </ForwardRef(Paper)>
          </WithStyles(ForwardRef(Paper))>
        </ForwardRef(AppBar)>
      </WithStyles(ForwardRef(AppBar))>
    </Component>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "space-between",
          "margin": ".5em 0 .6em 0",
          "maxWidth": 500,
        }
      }
    >
      <WithStyles(ForwardRef(Typography))
        align="center"
        color="textSecondary"
        gutterBottom={true}
        variant="subtitle1"
      >
        <ForwardRef(Typography)
          align="center"
          classes={
            Object {
              "alignCenter": "MuiTypography-alignCenter",
              "alignJustify": "MuiTypography-alignJustify",
              "alignLeft": "MuiTypography-alignLeft",
              "alignRight": "MuiTypography-alignRight",
              "body1": "MuiTypography-body1",
              "body2": "MuiTypography-body2",
              "button": "MuiTypography-button",
              "caption": "MuiTypography-caption",
              "colorError": "MuiTypography-colorError",
              "colorInherit": "MuiTypography-colorInherit",
              "colorPrimary": "MuiTypography-colorPrimary",
              "colorSecondary": "MuiTypography-colorSecondary",
              "colorTextPrimary": "MuiTypography-colorTextPrimary",
              "colorTextSecondary": "MuiTypography-colorTextSecondary",
              "displayBlock": "MuiTypography-displayBlock",
              "displayInline": "MuiTypography-displayInline",
              "gutterBottom": "MuiTypography-gutterBottom",
              "h1": "MuiTypography-h1",
              "h2": "MuiTypography-h2",
              "h3": "MuiTypography-h3",
              "h4": "MuiTypography-h4",
              "h5": "MuiTypography-h5",
              "h6": "MuiTypography-h6",
              "noWrap": "MuiTypography-noWrap",
              "overline": "MuiTypography-overline",
              "paragraph": "MuiTypography-paragraph",
              "root": "MuiTypography-root",
              "srOnly": "MuiTypography-srOnly",
              "subtitle1": "MuiTypography-subtitle1",
              "subtitle2": "MuiTypography-subtitle2",
            }
          }
          color="textSecondary"
          gutterBottom={true}
          variant="subtitle1"
        >
          <h6
            className="MuiTypography-root MuiTypography-subtitle1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom MuiTypography-alignCenter"
          >
            The preview images are not generated yet. Do you want to generate them?
          </h6>
        </ForwardRef(Typography)>
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Button))
        disabled={true}
        onClick={[Function]}
        variant="contained"
      >
        <ForwardRef(Button)
          classes={
            Object {
              "colorInherit": "MuiButton-colorInherit",
              "contained": "MuiButton-contained",
              "containedPrimary": "MuiButton-containedPrimary",
              "containedSecondary": "MuiButton-containedSecondary",
              "containedSizeLarge": "MuiButton-containedSizeLarge",
              "containedSizeSmall": "MuiButton-containedSizeSmall",
              "disableElevation": "MuiButton-disableElevation",
              "disabled": "Mui-disabled",
              "endIcon": "MuiButton-endIcon",
              "focusVisible": "Mui-focusVisible",
              "fullWidth": "MuiButton-fullWidth",
              "iconSizeLarge": "MuiButton-iconSizeLarge",
              "iconSizeMedium": "MuiButton-iconSizeMedium",
              "iconSizeSmall": "MuiButton-iconSizeSmall",
              "label": "MuiButton-label",
              "outlined": "MuiButton-outlined",
              "outlinedPrimary": "MuiButton-outlinedPrimary",
              "outlinedSecondary": "MuiButton-outlinedSecondary",
              "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
              "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
              "root": "MuiButton-root",
              "sizeLarge": "MuiButton-sizeLarge",
              "sizeSmall": "MuiButton-sizeSmall",
              "startIcon": "MuiButton-startIcon",
              "text": "MuiButton-text",
              "textPrimary": "MuiButton-textPrimary",
              "textSecondary": "MuiButton-textSecondary",
              "textSizeLarge": "MuiButton-textSizeLarge",
              "textSizeSmall": "MuiButton-textSizeSmall",
            }
          }
          disabled={true}
          onClick={[Function]}
          variant="contained"
        >
          <WithStyles(ForwardRef(ButtonBase))
            className="MuiButton-root MuiButton-contained Mui-disabled"
            component="button"
            disabled={true}
            focusRipple={true}
            focusVisibleClassName="Mui-focusVisible"
            onClick={[Function]}
            type="button"
          >
            <ForwardRef(ButtonBase)
              className="MuiButton-root MuiButton-contained Mui-disabled"
              classes={
                Object {
                  "disabled": "Mui-disabled",
                  "focusVisible": "Mui-focusVisible",
                  "root": "MuiButtonBase-root",
                }
              }
              component="button"
              disabled={true}
              focusRipple={true}
              focusVisibleClassName="Mui-focusVisible"
              onClick={[Function]}
              type="button"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-contained Mui-disabled Mui-disabled"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  <WithStyles(ForwardRef(CircularProgress))
                    size={24}
                    style={
                      Object {
                        "marginRight": "1em",
                      }
                    }
                  >
                    <ForwardRef(CircularProgress)
                      classes={
                        Object {
                          "circle": "MuiCircularProgress-circle",
                          "circleDisableShrink": "MuiCircularProgress-circleDisableShrink",
                          "circleIndeterminate": "MuiCircularProgress-circleIndeterminate",
                          "circleStatic": "MuiCircularProgress-circleStatic",
                          "colorPrimary": "MuiCircularProgress-colorPrimary",
                          "colorSecondary": "MuiCircularProgress-colorSecondary",
                          "indeterminate": "MuiCircularProgress-indeterminate",
                          "root": "MuiCircularProgress-root",
                          "static": "MuiCircularProgress-static",
                          "svg": "MuiCircularProgress-svg",
                        }
                      }
                      size={24}
                      style={
                        Object {
                          "marginRight": "1em",
                        }
                      }
                    >
                      <div
                        className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                        role="progressbar"
                        style={
                          Object {
                            "height": 24,
                            "marginRight": "1em",
                            "width": 24,
                          }
                        }
                      >
                        <svg
                          className="MuiCircularProgress-svg"
                          viewBox="22 22 44 44"
                        >
                          <circle
                            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                            cx={44}
                            cy={44}
                            fill="none"
                            r={20.2}
                            strokeWidth={3.6}
                            style={Object {}}
                          />
                        </svg>
                      </div>
                    </ForwardRef(CircularProgress)>
                  </WithStyles(ForwardRef(CircularProgress))>
                  Generate
                </span>
              </button>
            </ForwardRef(ButtonBase)>
          </WithStyles(ForwardRef(ButtonBase))>
        </ForwardRef(Button)>
      </WithStyles(ForwardRef(Button))>
    </div>
  </div>
</Component>
`;

exports[`Document Regenerate Previews component should render without crashing 1`] = `
<div
  style={
    Object {
      "alignContent": "center",
      "alignItems": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "space-around",
    }
  }
>
  <Component
    style={
      Object {
        "position": "fixed",
        "top": 0,
      }
    }
  >
    <span />
  </Component>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
        "margin": ".5em 0 .6em 0",
        "maxWidth": 500,
      }
    }
  >
    <WithStyles(ForwardRef(Typography))
      align="center"
      color="textSecondary"
      gutterBottom={true}
      variant="subtitle1"
    >
      The preview images are not generated yet. Do you want to generate them?
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Button))
      disabled={false}
      onClick={[Function]}
      variant="contained"
    >
      Generate
    </WithStyles(ForwardRef(Button))>
  </div>
</div>
`;
